### PR TITLE
`Exam mode`: Allow to explicitly terminate exam participation without submitting

### LINKED ISSUE #7107: `Exam`: Allow to explicitly terminate exam participation without submitting
### Is your feature request related to a problem?

When participating in an exam supervised in a lecture hall, students may decide to withdraw early from the exam participation. In case the student submits the exam, the supervisor in the lecture hall can check that the student has  submitted their exam. However, in case the student says that they do not want the exam to be graded, and still want to leave early, their currently is no way for the student to terminate the exam, which would allow the superivsor to ensure that the student can no longer work on the exam after the students has left the room. 
 

### Describe the solution you'd like

There should be an explicit option to chose to terminate their exam participation without submitting the exam for grading, so that the supervisor can see verify that the exam participation has ended.
This option would correspond to a student to strike through their exam in a pen & paper exam before handing in.

### Describe alternatives you've considered

One could think of just enforce that students who submit early to submit the exam as usual. If they want to avoid passing the exam with a bad grade, they would have to empty all their responses to exercises. However, in single-choice exercises, it is not possible AFAIK to not submit anything if you already participated in the single-choice exercise.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by valentin-boehm**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://ls1intum.github.io/Artemis/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://ls1intum.github.io/Artemis/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Closes #7107. Students might want to withdraw and leave early from an exam without submitting. Currently, this is not possible because the student could continue working on the exam after leaving the lecture hall. There is no way for supervisors to make sure that the student cannot submit after leaving.

### Description

I added a flag `abandoned` to StudentExam. Abandoned student exams cannot be submitted (and vice-versa).
Students can abandon their exam through a button in the hand-in-early dialog. They need to confirm their decision (similar to handing-in early). Instructors can see that a student abandoned the exam in the student exam overview. Additionally, they see the total number of abandons in the exam status and the exam checklist.
After abandoning, it is impossible for the student to submit the exam, so supervisors can be sure that the student cannot continue working after leaving.
During assessment, abandoned exams will be treated like unsubmitted exams (because they _are_ unsubmitted).

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Active exam

As student:
1. Participate in the exam
2. Click on "Hand in early" and then on "Abandon exam"
3. In the confirmation modal, click on "Back"
4. Make sure that you come back to hand-in-early dialog
5. Click on "Abandon" again
6. Make sure that the confirmation button "Abandon exam" is only enabled when you check the box and enter your name correctly
7. Click the confirmation button
8. Make sure that you cannot access the exam anymore and instead see the information that you abandoned the exam

As instructor:
1. Make sure that you see the correct number of abandoned exams (1) in the exam status and the exam checklist
2. Make sure that you can see that the student abandoned in the student exam list and in the detail view of the student exam
3. In the detail view of the student exam, make sure that it is not submitted
4. After the time is over: Make sure that the button "Change student exam to submitted" is disabled for abandoned exams

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [exam-checklist.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5596/latest/artifact/shared/Coverage-Report-Client-Tests/app/exam/manage/exams/exam-checklist-component/exam-checklist.service.ts.html) | 100% | ✅ |
| [exam-navigation-bar.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5596/latest/artifact/shared/Coverage-Report-Client-Tests/app/exam/participate/exam-navigation-bar/exam-navigation-bar.component.ts.html) | 87.62% | ✅  |
| [exam-participation.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5596/latest/artifact/shared/Coverage-Report-Client-Tests/app/exam/participate/exam-participation.component.ts.html) | 89.83% | ✅  |
| [exam-participation.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5596/latest/artifact/shared/Coverage-Report-Client-Tests/app/exam/participate/exam-participation.service.ts.html) | 77% | ✅  |
| [exam-termination-confirmation.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5596/latest/artifact/shared/Coverage-Report-Client-Tests/app/exam/participate/exam-termination-confirmation/exam-termination-confirmation.component.ts.html) | 100% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [StudentExamRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5596/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository/StudentExamRepository.html) | 96% | ✅  |
| [StudentExamService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5596/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.exam/StudentExamService.html) | 94% | ✅  |
| [StudentExamResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5596/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/StudentExamResource.html) | 87% | ✅  |
| [ExamChecklistDTO.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5596/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.dto/ExamChecklistDTO.html) | 100% | ✅ |
| [StudentExam.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5596/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.exam/StudentExam.html) | 98% | ✅  |

### Screenshots

Bildschirmfoto 2023-11-17 um 15 26 21

Abandon-button in the hand-in-early dialog

Bildschirmfoto 2023-10-18 um 15 14 43

Confirmation modal for abandoning

Bildschirmfoto 2023-10-18 um 15 15 11

After abandoning, the student cannot access the exam anymore

## Summary by CodeRabbit

- **New Features**
  - Introduced an "abandon exam" feature allowing students to indicate if they have abandoned the exam.
  - Exam status now displays the number of abandoned exams and the percentage of total exams.
  - Added visual indicators for abandoned exams in various exam management components.

- **Enhancements**
  - Improved exam checklist to account for abandoned exams when calculating progress.
  - Enhanced student exam detail views with abandonment status information.
  - Added the ability to sort by "Abandoned" status in exam management tables.

- **Bug Fixes**
  - Corrected exam submission logic to handle cases where exams are abandoned.
  - Updated test run management to reflect abandoned exam status accurately.

- **User Interface**
  - Implemented a confirmation dialog for exam abandonment.
  - Updated exam participation interfaces to support exam abandonment actions.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/domain/exam/StudentExam.java
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/domain/exam/StudentExam.java:29]:
```suggestion
    private boolean abandoned = false;
```
Since the database uses a default, this can be used here as well to avoid having to deal with `Booleans.TRUE.equals(abondoned)` and similar.

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentExamResource.java
Comment (valentin-boehm) [reply src/main/java/de/tum/in/www1/artemis/web/rest/StudentExamResource.java:313]: Yes, this check is used for most student requests in this resource. StudentExams are only generated for registered students and deleted when a student unregisters, so it should not be possible to have a studentExam for an unregistered student.

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentExamResource.java
Comment (valentin-boehm) [reply src/main/java/de/tum/in/www1/artemis/web/rest/StudentExamResource.java:329]: Thanks, I will change that 👍

**Review by Strohgelaender — DISMISSED** [review]
code lgtm

**Review by JohannesStoehr — APPROVED** [review]
Code looks good to me. Left one optional question about the logging

File: src/main/java/de/tum/in/www1/artemis/service/exam/StudentExamService.java
Comment (valentin-boehm) [reply src/main/java/de/tum/in/www1/artemis/service/exam/StudentExamService.java]: You're right, with the LogginAspect this is kind of redundant. I removed it

**Review by Strohgelaender — APPROVED** [review]
code

**Review by JohannesStoehr — DISMISSED** [review]
Reapprove code

**Comment by farisd16** [comment]
Tested on ts3, works as expected

**Review by farisd16 — DISMISSED** [review]
Tested on ts3, works as expected

**Review by JohannesStoehr — DISMISSED** [review]
Code after merge

File: src/main/java/de/tum/in/www1/artemis/domain/exam/StudentExam.java
Comment (tobias-lippert) [inline src/main/java/de/tum/in/www1/artemis/domain/exam/StudentExam.java:74]: Why does this return the non-primitive type?

File: src/main/java/de/tum/in/www1/artemis/domain/exam/StudentExam.java
Comment (tobias-lippert) [inline src/main/java/de/tum/in/www1/artemis/domain/exam/StudentExam.java:77]: We shouldn't set a Boolean to a boolean.

**Review by tobias-lippert — DISMISSED** [review]

**Review by JohannesStoehr — APPROVED** [review]
Code

File: src/main/resources/config/liquibase/changelog/20231016123944_changelog.xml
Comment (Hialus) [inline src/main/resources/config/liquibase/changelog/20231016123944_changelog.xml:10]: Note for your future PRs: Since there is a clear default value for this attribute both logically and in the Java code, you could also specify it here. Would also prevent the DB from breaking exams on other PRs 😉

**Review by Hialus — APPROVED** [review]
Code looks good to me 👍🏻

**Review by AntonGeTUM — COMMENTED** [review]
Tested on leg. TS3, the feature itself worked as expected but in both Chrome and Firefox and on two different PCs the checkbox for abandoning the exam appeared as a thin red line that turned green when clicking it, see picture below.
image

**Comment by valentin-boehm** [comment]
> Tested on leg. TS3, the feature itself worked as expected but in both Chrome and Firefox and on two different PCs the checkbox for abandoning the exam appeared as a thin red line that turned green when clicking it, see picture below.
@AntonGeTUM Do you have the same issue with other checkboxes, e.g. when handing in early?

**Comment by AntonGeTUM** [comment]
> Do you have the same issue with other checkboxes, e.g. when handing in early?

No, the checkbox before beginning the exam and the one for handing early were both normal.

**Comment by MaximilianJG** [comment]
Tested on Artemis server 2 in a testing session with @Ceiridge (who was an instructor). 

Works as expected! Please see screenshots: 

Screenshot 2023-11-15 at 17 31 43
Screenshot 2023-11-15 at 17 31 56

**Comment by Ceiridge** [comment]
Tested with @MaximilianJG in a testing session on test2. Works as expected

image
image
image

**Review by Gusti2010 — APPROVED** [review]
tested manually on TS11_Legacy. Everything works as expected!

**Comment by valentin-boehm** [comment]
I moved the Abandon button into the hand-in-early dialog to make it less prominent, since it's not something students will need regularly.

**Comment by mateusmm01** [comment]
Tested on ts1, works as expected also with more than one student abandoning the exam.

**Review by farisd16 — APPROVED** [review]
Tested on TS1, works as expected

**Review by Strohgelaender — APPROVED** [review]
re-approve code

**Review by lennart-keller — DISMISSED** [review]
Code lgtm

**Review by mateusmm01 — DISMISSED** [review]
Reapprove!

**Comment by Ceiridge** [comment]
Works as expected on ts1

image
image

The button is also disabled, as expected.

image

**Review by rabeatwork — CHANGES_REQUESTED** [review]

**Review by mateusmm01 — DISMISSED** [review]
Retested on ts1. Works as expected!

**Review by lennart-keller — APPROVED** [review]
Reapprove code after merge

**Review by farisd16 — APPROVED** [review]
Reapprove after fixing merge conflict

**Review by mateusmm01 — CHANGES_REQUESTED** [review]
should have been approval…

**Review by mateusmm01 — APPROVED** [review]
Reapprove after changes

File: src/test/javascript/spec/component/exam/participate/exam-participation.component.spec.ts
Comment (valentin-boehm) [reply src/test/javascript/spec/component/exam/participate/exam-participation.component.spec.ts:698]: Since we explicitly set abandoned to false this is not necessary

File: src/test/java/de/tum/in/www1/artemis/exam/StudentExamIntegrationTest.java
Comment (valentin-boehm) [reply src/test/java/de/tum/in/www1/artemis/exam/StudentExamIntegrationTest.java:2831]: The response status is checked.

File: src/test/java/de/tum/in/www1/artemis/exam/StudentExamIntegrationTest.java
Comment (valentin-boehm) [reply src/test/java/de/tum/in/www1/artemis/exam/StudentExamIntegrationTest.java:2821]: The response status is checked.

File: src/test/java/de/tum/in/www1/artemis/exam/StudentExamIntegrationTest.java
Comment (valentin-boehm) [reply src/test/java/de/tum/in/www1/artemis/exam/StudentExamIntegrationTest.java:2810]: The response status is checked.

File: src/test/java/de/tum/in/www1/artemis/exam/StudentExamIntegrationTest.java
Comment (valentin-boehm) [reply src/test/java/de/tum/in/www1/artemis/exam/StudentExamIntegrationTest.java:1063]: The response status is checked.

File: src/main/java/de/tum/in/www1/artemis/service/exam/StudentExamService.java
Comment (valentin-boehm) [reply src/main/java/de/tum/in/www1/artemis/service/exam/StudentExamService.java:303]: studentExam should never be null at this point. If it unexpectedly is null, that would indicate a serious issue in other parts of the code and throwing a NullPointer is ok in that scenario.

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentExamResource.java
Comment (valentin-boehm) [reply src/main/java/de/tum/in/www1/artemis/web/rest/StudentExamResource.java:332]: This additional info is already include in the error logging.

File: src/test/java/de/tum/in/www1/artemis/exam/StudentExamIntegrationTest.java
Comment (valentin-boehm) [reply src/test/java/de/tum/in/www1/artemis/exam/StudentExamIntegrationTest.java:2857]: We are already checking that the exam is not abandoned after the forbidden operation.

File: src/test/java/de/tum/in/www1/artemis/exam/StudentExamIntegrationTest.java
Comment (valentin-boehm) [reply src/test/java/de/tum/in/www1/artemis/exam/StudentExamIntegrationTest.java:2846]: Since the control flow of the respective method in the StudentExamRessource is very simple, we can be certain that the studentExam has not changed when we receive the HttpStatus `forbidden` and it is not necessary to include an additional check here.

File: src/test/java/de/tum/in/www1/artemis/exam/StudentExamIntegrationTest.java
Comment (valentin-boehm) [reply src/test/java/de/tum/in/www1/artemis/exam/StudentExamIntegrationTest.java:993]: Since the control flow of the respective method in the StudentExamRessource is very simple, we can be certain that the studentExam has not changed when we receive the HttpStatus forbidden and it is not necessary to include an additional check here.

File: src/test/java/de/tum/in/www1/artemis/exam/StudentExamIntegrationTest.java
Comment (valentin-boehm) [reply src/test/java/de/tum/in/www1/artemis/exam/StudentExamIntegrationTest.java:980]: Since the control flow of the respective method in the StudentExamRessource is very simple, we can be certain that the studentExam has not changed when we receive the HttpStatus forbidden and it is not necessary to include an additional check here.

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentExamResource.java
Comment (valentin-boehm) [reply src/main/java/de/tum/in/www1/artemis/web/rest/StudentExamResource.java:892]: Since the cause of this exception is self-explanatory, I would argue that it is not necessary to include a detailed error message here.

**Review by farisd16 — APPROVED** [review]
Reapprove

**Review by georgyia — APPROVED** [review]
tested manually on ts1, works as expected

**Review by lennart-keller — APPROVED** [review]
Reapprove code

**Review by mateusmm01 — APPROVED** [review]
Reapprove after changes

**Pull request closed without merging**
